File: Bit Manipulation/M-318. Maximum Product of Word Lengths/M-318. Maximum Product of Word Lengths.ts
function maxProduct(words: string[]): number {
  let result = 0;
  const n = words.length;
  const masks: number[] = new Array(n);
  const unicodeOfA = 'a'.charCodeAt(0);

  for (let i = 0; i < n; i++) {
    let mask = 0;
    for (let j = 0; j < words[i].length; j++) {
      const unicode = words[i].charCodeAt(j) - unicodeOfA;
      mask |= 1 << unicode;
    }
    masks[i] = mask;
  }

  for (let i = 0; i < n - 1; i++) {
    for (let j = 0; j < n; j++) {
      if ((masks[i] & masks[j]) === 0) {
        result = Math.max(result, words[i].length * words[j].length);
      }
    }
  }

  return result;
}

const testData = [
  { words: ['abcw', 'baz', 'foo', 'bar', 'xtfn', 'abcdef'], ans: 16 },
  { words: ['a', 'ab', 'abc', 'd', 'cd', 'bcd', 'abcd'], ans: 4 },
  { words: ['a', 'aa', 'aaa', 'aaaa'], ans: 0 },
];

for (const data of testData) {
  console.log(data.ans === maxProduct(data.words));
}

File: Bit Manipulation/M-318. Maximum Product of Word Lengths/M-318. Maximum Product of Word Lengths.py
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        n = len(words)
        result = 0
        unicode_of_a = ord("a")
        masks = [None] * n

        for i in range(n):
            mask = 0
            for j in range(len(words[i])):
                unicode = ord(words[i][j]) - unicode_of_a
                mask |= 1 << unicode
            masks[i] = mask

        for i in range(n - 1):
            for j in range(n):
                if masks[i] & masks[j] == 0:
                    result = max(result, len(words[i]) * len(words[j]))

        return result
